hubstaff-export.rb

```diff
@@ -32,29 +33,275 @@
+
+  def load
+    return unless File.exist?(config_file)
+
+    @config = JSON.parse(File.read(config_file), symbolize_names: true)
+  rescue
+    @config = {}
+  end
+
+  def save!
+    File.open(config_file, 'w') do |file|
+      file.write(JSON.generate(@config, {indent: '  ', object_nl: "\n", array_nl: "\n"}))
+    end
+  end
+
+  def [](key)
+    @config[key]
+  end
+
+  def []=(key, value)
+    @config[key] = value
+  end
+
+  def delete(key)
+    @config.delete(key)
+  end
+
+  def dig(*keys)
+    @config.dig(*keys)
+  end
+end
+
+# Class to manage the token and triggering automatic refeshes
+class TokenStore
+  # @param auth_client [AuthClient]
+  # @param config_store [ConfigStore]
+  def initialize(auth_client, config_store)
+    @auth_client  = auth_client
+    @config_store = config_store
+  end
+
+  # @return [Hash]
+  def headers
+    {
+      Authorization: "Bearer #{access_token}",
+    }
+  end
+
+  # @return [String]
+  def access_token
+    token_data = decode_token(@config_store.dig(:token, :access_token))
+    @auth_client.refresh_token! if token_data[:exp] < (Time.now.to_i + 300)
+
+    raise TokenError, "No access token" if @config_store.dig(:token, :access_token).nil?
+
+    @config_store.dig(:token, :access_token)
+  end
+
+  def decode_token(token)
+    return {exp: 0} if token.nil?
+
+    parts = token.split('.')
+    JSON.parse(Base64.decode64(parts[1]), symbolize_names: true)
+  rescue
+    {exp: 0}
   end
 end
 
+# Simple Hubstaff API Wrapper
+class APIClient
+  # @param base_url [String]
+  # @param token_store [TokenStore]
+  # @param skip_ssl_verify [Boolean]
+  def initialize(base_url, token_store, skip_ssl_verify: false)
+    @http_client = HTTPClient.new(skip_ssl_verify: skip_ssl_verify)
+
+    @base_url    = URI.parse(base_url)
+    @token_store = token_store
+  end
+
+  def post(path, params)
+    @http_client.post_json(@base_url.merge(path), params: params, headers: @token_store.headers)
+  end
+
+  def get(path, params)
+    @http_client.get(@base_url.merge(path), params: params, headers: @token_store.headers)
+  end
+end
+
+# Main class for the export tool
 class HubstaffExport
-  VERSION = '0.5.0'
+  VERSION = '0.6.0'
+
+  CONFIG_FILE = 'hubstaff-client.cfg'
 
   attr_reader :options
+  attr_reader :config_store
+  attr_reader :auth_client
+  attr_reader :token_store
+  attr_reader :api_client
 
   def initialize(arguments)
     @arguments = arguments
     # Set defaults
-    @options = OpenStruct.new
-    @options.verbose = false
-    @options.image_format = 'full'
-    @options.directory = 'screens'
+    @options                 = OpenStruct.new
+    @options.verbose         = false
+    @options.image_format    = 'full'
+    @options.directory       = 'screens'
     @options.skip_ssl_verify = false
-    @api_url = 'https://api.hubstaff.com/v1'
+    @discovery_url           = 'https://account.hubstaff.com/.well-known/openid-configuration'
+    @api_url                 = 'https://api.hubstaff.com/v2/'
```

Comment:
`@discovery_url` and `@api_url` look like good candidates to be extracted into a constant.

Reply:
yeah..  switched